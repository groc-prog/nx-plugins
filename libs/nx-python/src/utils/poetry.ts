import type { SpawnSyncOptions } from 'child_process';

import chalk from 'chalk';
import spawn from 'cross-spawn';
import commandExists from 'command-exists';

export type PyProjectTomlDependency =
  | string
  | {
      path?: string;
      develop?: boolean;
    };

export type PyProjectTomlDependencies = {
  [key: string]: PyProjectTomlDependency;
};

export type PyProjectToml = {
  tool?: {
    poetry?: {
      name: string;
      version: string;
      packages?: Array<{
        include: string;
      }>;
      dependencies?: PyProjectTomlDependencies;
      group?: {
        [key: string]: {
          dependencies: PyProjectTomlDependencies;
        };
      };
    };
  };
};

export const POETRY_EXECUTABLE = 'poetry';

/**
 * Checks if Poetry is installed.
 *
 * @throws {Error} If Poetry is not installed.
 */
export async function checkPoetryExecutable(): Promise<void> {
  try {
    await commandExists(POETRY_EXECUTABLE);
  } catch (e) {
    throw new Error('Poetry is not installed. Please install Poetry before running this command.');
  }
}

/**
 * Run poetry with given arguments.
 *
 * @param {string[]} args - Arguments to pass to poetry.
 * @param {SpawnSyncOptions} options - Options to pass to spawn.
 */
export function runPoetry(args: string[], options: SpawnSyncOptions = {}): void {
  const commandStr = `${POETRY_EXECUTABLE} ${args.join(' ')}`;
  console.log(chalk`Running command {bold ${commandStr}}`);

  const result = spawn.sync(POETRY_EXECUTABLE, args, {
    ...options,
    shell: false,
    stdio: 'inherit',
    maxBuffer: 1024 * 1024 * 10,
  });

  if (result.status !== 0)
    throw new Error(chalk`{bold ${commandStr}} command failed with exit code {bold ${result.status}}`);
}

// export function checkSharedVenv(workspaceRoot: string): void {
//   const workspacePyProjectToml = path.join(workspaceRoot, 'pyproject.toml');

//   if (existsSync(workspacePyProjectToml)) {

//   }
// }
// export function activateVenv(workspaceRoot: string) {
//   if (!process.env.VIRTUAL_ENV) {
//     const rootPyproject = path.join(workspaceRoot, 'pyproject.toml');

//     if (fs.existsSync(rootPyproject)) {
//       const rootConfig = parse(fs.readFileSync(rootPyproject, 'utf-8')) as PyProjectToml;
//       const autoActivate = rootConfig.tool.nx?.autoActivate ?? false;
//       if (autoActivate) {
//         console.log(chalk`\n{bold shared virtual environment detected and not activated, activating...}\n\n`);
//         const virtualEnv = path.resolve(workspaceRoot, '.venv');
//         process.env.VIRTUAL_ENV = virtualEnv;
//         process.env.PATH = `${virtualEnv}/bin:${process.env.PATH}`;
//         delete process.env.PYTHONHOME;
//       }
//     }
//   }
// }
