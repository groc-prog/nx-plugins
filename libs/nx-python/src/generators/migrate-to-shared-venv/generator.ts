import {
  readJsonFile,
  formatFiles,
  generateFiles,
  getProjects,
  Tree,
  ProjectConfiguration,
} from '@nx/devkit';
import path from 'path';
import { Schema } from './schema';
import { parse, stringify } from '@iarna/toml';
import { PyprojectToml } from '../../graph/dependency-graph';
import chalk from 'chalk';
import { checkPoetryExecutable, runPoetry } from '../../executors/utils/poetry';

async function addFiles(host: Tree, options: Schema) {
  const packageJson = await readJsonFile('package.json');

  const templateOptions = {
    ...options,
    template: '',
    dot: '.',
    packageName: packageJson.name,
    description: packageJson.description || '',
  };

  generateFiles(host, path.join(__dirname, 'files'), '.', templateOptions);
}

type LockUpdateTask = () => void;

function updatePyprojectRoot(host: Tree, options: Schema): LockUpdateTask[] {
  const postGeneratorTasks = [];

  const rootPyprojectToml = parse(
    host.read('pyproject.toml').toString()
  ) as PyprojectToml;

  for (const project of getProjects(host)) {
    const [, projectConfig] = project;
    const pyprojectTomlPath = path.join(projectConfig.root, 'pyproject.toml');
    if (host.exists(pyprojectTomlPath)) {
      const pyprojectToml = parse(
        host.read(pyprojectTomlPath).toString()
      ) as PyprojectToml;

      rootPyprojectToml.tool.poetry.dependencies[
        pyprojectToml.tool.poetry.name
      ] = {
        path: projectConfig.root,
        develop: true,
      };
      if (options.moveDevDependencies) {
        postGeneratorTasks.push(
          moveDevDependencies(
            pyprojectToml,
            rootPyprojectToml,
            host,
            pyprojectTomlPath,
            projectConfig
          )
        );
      }
    }
  }

  host.write('pyproject.toml', stringify(rootPyprojectToml));

  return postGeneratorTasks;
}

function moveDevDependencies(
  pyprojectToml: PyprojectToml,
  rootPyprojectToml: PyprojectToml,
  host: Tree,
  pyprojectTomlPath: string,
  projectConfig: ProjectConfiguration
) {
  const devDependencies =
    pyprojectToml.tool.poetry.group?.dev?.dependencies || {};

  for (const devDependency of Object.keys(devDependencies)) {
    rootPyprojectToml.tool.poetry.group =
      rootPyprojectToml.tool.poetry.group || {};
    rootPyprojectToml.tool.poetry.group.dev = rootPyprojectToml.tool.poetry
      .group.dev || { dependencies: {} };
    rootPyprojectToml.tool.poetry.group.dev.dependencies[devDependency] =
      devDependencies[devDependency];
  }

  if (Object.keys(devDependencies).length > 0) {
    delete pyprojectToml.tool.poetry.group.dev;
  }
  host.write(pyprojectTomlPath, stringify(pyprojectToml));

  return () => {
    console.log(
      chalk`  Updating ${pyprojectToml.tool.poetry.name} {bgBlue poetry.lock}...`
    );
    const lockArgs = ['lock', '--no-update'];
    runPoetry(lockArgs, { cwd: projectConfig.root, log: false });
    console.log(chalk`\n  {bgBlue poetry.lock} updated.\n`);
  };
}

function updateRootPoetryLock() {
  console.log(chalk`  Updating root {bgBlue poetry.lock}...`);
  runPoetry(['install'], { log: false });
  console.log(chalk`\n  {bgBlue poetry.lock} updated.\n`);
}

async function generator(host: Tree, options: Schema) {
  await checkPoetryExecutable();

  await addFiles(host, options);
  const lockUpdateTasks = updatePyprojectRoot(host, options);
  await formatFiles(host);

  return () => {
    lockUpdateTasks.forEach((task) => task());
    updateRootPoetryLock();
  };
}

export default generator;
