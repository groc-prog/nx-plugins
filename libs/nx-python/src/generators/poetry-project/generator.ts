import {
  addProjectConfiguration,
  formatFiles,
  generateFiles,
  getWorkspaceLayout,
  names,
  offsetFromRoot,
  ProjectConfiguration,
  readProjectConfiguration,
  Tree,
} from '@nx/devkit';
import * as path from 'path';
import { PoetryProjectGeneratorSchema } from './schema';
import { checkPoetryExecutable, runPoetry } from '../../executors/utils/poetry';
import {
  PyprojectToml,
  PyprojectTomlDependencies,
} from '../../graph/dependency-graph';
import { parse, stringify } from '@iarna/toml';
import chalk from 'chalk';
import _ from 'lodash';

interface NormalizedSchema extends PoetryProjectGeneratorSchema {
  projectName: string;
  projectRoot: string;
  projectDirectory: string;
  individualPackage: boolean;
  devDependenciesProjectPath?: string;
  devDependenciesProjectPkgName?: string;
  pythonAddopts?: string;
  parsedTags: string[];
}

function normalizeOptions(
  tree: Tree,
  options: PoetryProjectGeneratorSchema
): NormalizedSchema {
  const name = names(options.name).fileName;
  const projectDirectory = options.directory
    ? `${names(options.directory).fileName}/${name}`
    : name;
  const projectName = projectDirectory.replace(/\//g, '-');
  const projectRoot = `${
    options.projectType === 'application'
      ? getWorkspaceLayout(tree).appsDir
      : getWorkspaceLayout(tree).libsDir
  }/${projectDirectory}`;
  const parsedTags = options.tags
    ? options.tags.split(',').map((s) => s.trim())
    : [];

  const newOptions = _.clone(options) as NormalizedSchema;

  if (!options.pyprojectPythonDependency) {
    newOptions.pyprojectPythonDependency = '>=3.9,<3.11';
  }

  if (!options.pyenvPythonVersion) {
    newOptions.pyenvPythonVersion = '3.9.5';
  }

  if (!options.moduleName) {
    newOptions.moduleName = projectName.replace(/-/g, '_');
  }

  if (!options.packageName) {
    newOptions.packageName = projectName;
  }

  if (!options.description) {
    newOptions.description = 'Automatically generated by Nx.';
  }
  if (options.devDependenciesProject) {
    const projectConfig = readProjectConfiguration(
      tree,
      options.devDependenciesProject
    );
    newOptions.devDependenciesProjectPath = path.relative(
      projectRoot,
      projectConfig.root
    );
  }

  const pythonAddopts = getPyTestAddopts(options, projectRoot);

  if (options.unitTestRunner === 'none') {
    newOptions.unitTestHtmlReport = false;
    newOptions.unitTestJUnitReport = false;
    newOptions.codeCoverage = false;
    newOptions.codeCoverageHtmlReport = false;
    newOptions.codeCoverageXmlReport = false;
    newOptions.codeCoverageThreshold = undefined;
  }

  let devDependenciesProjectPkgName: string | undefined;
  if (options.devDependenciesProject) {
    const { pyprojectToml } = getPyprojectTomlByProjectName(
      tree,
      options.devDependenciesProject
    );
    devDependenciesProjectPkgName = pyprojectToml.tool.poetry.name;
  }

  return {
    ...options,
    ...newOptions,
    devDependenciesProject: options.devDependenciesProject || '',
    individualPackage: !tree.exists('pyproject.toml'),
    devDependenciesProjectPkgName,
    pythonAddopts,
    projectName,
    projectRoot,
    projectDirectory,
    parsedTags,
  };
}

function getPyTestAddopts(
  options: PoetryProjectGeneratorSchema,
  projectRoot: string
): string | undefined {
  if (options.unitTestRunner === 'pytest') {
    const args = [];
    const offset = offsetFromRoot(projectRoot);
    if (options.codeCoverage) {
      args.push('--cov');
    }
    if (options.codeCoverageThreshold) {
      args.push(`--cov-fail-under=${options.codeCoverageThreshold}`);
    }
    if (options.codeCoverage && options.codeCoverageHtmlReport) {
      args.push(`--cov-report html:'${offset}coverage/${projectRoot}/html'`);
    }

    if (options.codeCoverage && options.codeCoverageXmlReport) {
      args.push(
        `--cov-report xml:'${offset}coverage/${projectRoot}/coverage.xml'`
      );
    }

    if (options.unitTestHtmlReport) {
      args.push(
        `--html='${offset}reports/${projectRoot}/unittests/html/index.html'`
      );
    }

    if (options.unitTestJUnitReport) {
      args.push(
        `--junitxml='${offset}reports/${projectRoot}/unittests/junit.xml'`
      );
    }

    return args.join(' ');
  }
}

function addFiles(tree: Tree, options: NormalizedSchema) {
  const templateOptions = {
    ...options,
    ...names(options.name),
    offsetFromRoot: offsetFromRoot(options.projectRoot),
    template: '',
    dot: '.',
  };
  if (options.templateDir) {
    generateFiles(
      tree,
      path.join(options.templateDir),
      options.projectRoot,
      templateOptions
    );
    return;
  }

  generateFiles(
    tree,
    path.join(__dirname, 'files', 'base'),
    options.projectRoot,
    templateOptions
  );

  if (options.unitTestRunner === 'pytest') {
    generateFiles(
      tree,
      path.join(__dirname, 'files', 'pytest'),
      options.projectRoot,
      templateOptions
    );
  }

  if (options.linter === 'flake8') {
    generateFiles(
      tree,
      path.join(__dirname, 'files', 'flake8'),
      options.projectRoot,
      templateOptions
    );
  }
}

function updateRootPyprojectToml(
  host: Tree,
  normalizedOptions: NormalizedSchema
) {
  if (!normalizedOptions.individualPackage) {
    const rootPyprojectToml = parse(
      host.read('./pyproject.toml', 'utf-8')
    ) as PyprojectToml;

    const group = normalizedOptions.rootPyprojectDependencyGroup ?? 'main';

    if (group === 'main') {
      rootPyprojectToml.tool.poetry.dependencies[
        normalizedOptions.packageName
      ] = {
        path: normalizedOptions.projectRoot,
        develop: true,
      };
    } else {
      rootPyprojectToml.tool.poetry.group = {
        ...(rootPyprojectToml.tool.poetry.group || {}),
        [group]: {
          ...(rootPyprojectToml.tool.poetry.group?.[group] || {}),
          dependencies: {
            ...(rootPyprojectToml.tool.poetry.group?.[group]?.dependencies ||
              {}),
            [normalizedOptions.packageName]: {
              path: normalizedOptions.projectRoot,
              develop: true,
            },
          },
        },
      };
    }

    if (!normalizedOptions.devDependenciesProject) {
      const { changed, dependencies } = addTestDependencies(
        rootPyprojectToml.tool.poetry.group?.dev?.dependencies || {},
        normalizedOptions
      );

      if (changed) {
        rootPyprojectToml.tool.poetry.group = {
          ...(rootPyprojectToml.tool.poetry.group || {}),
          dev: {
            dependencies: dependencies,
          },
        };
      }
    }

    host.write('./pyproject.toml', stringify(rootPyprojectToml));
  }
}

function updateDevDependenciesProject(
  host: Tree,
  normalizedOptions: NormalizedSchema
) {
  if (normalizedOptions.devDependenciesProject) {
    const { pyprojectToml, pyprojectTomlPath } = getPyprojectTomlByProjectName(
      host,
      normalizedOptions.devDependenciesProject
    );

    const { changed, dependencies } = addTestDependencies(
      pyprojectToml.tool.poetry.dependencies,
      normalizedOptions
    );

    if (changed) {
      pyprojectToml.tool.poetry.dependencies = {
        ...pyprojectToml.tool.poetry.dependencies,
        ...dependencies,
      };

      host.write(pyprojectTomlPath, stringify(pyprojectToml));
    }
  }
}

function getPyprojectTomlByProjectName(host: Tree, projectName: string) {
  const projectConfig = readProjectConfiguration(host, projectName);
  const pyprojectTomlPath = path.join(projectConfig.root, 'pyproject.toml');

  const pyprojectToml = parse(
    host.read(pyprojectTomlPath, 'utf-8')
  ) as PyprojectToml;

  return { pyprojectToml, pyprojectTomlPath };
}

function addTestDependencies(
  dependencies: PyprojectTomlDependencies,
  normalizedOptions: NormalizedSchema
) {
  const originalDependencies = _.clone(dependencies);

  if (normalizedOptions.linter === 'flake8' && !dependencies['flake8']) {
    dependencies['flake8'] = '6.0.0';
  }

  if (!dependencies['autopep8']) {
    dependencies['autopep8'] = '2.0.2';
  }

  if (
    normalizedOptions.unitTestRunner === 'pytest' &&
    !dependencies['pytest']
  ) {
    dependencies['pytest'] = '7.3.1';
  }
  if (
    normalizedOptions.unitTestRunner === 'pytest' &&
    !dependencies['pytest-sugar']
  ) {
    dependencies['pytest-sugar'] = '0.9.7';
  }

  if (
    normalizedOptions.unitTestRunner === 'pytest' &&
    normalizedOptions.codeCoverage &&
    !dependencies['pytest-cov']
  ) {
    dependencies['pytest-cov'] = '4.1.0';
  }

  if (
    normalizedOptions.unitTestRunner === 'pytest' &&
    normalizedOptions.codeCoverageHtmlReport &&
    !dependencies['pytest-html']
  ) {
    dependencies['pytest-html'] = '3.2.0';
  }

  return {
    changed: !_.isEqual(originalDependencies, dependencies),
    dependencies,
  };
}

function updateRootPoetryLock(host: Tree) {
  if (host.exists('./pyproject.toml')) {
    console.log(chalk`  Updating root {bgBlue poetry.lock}...`);
    runPoetry(['lock', '--no-update'], { log: false });
    runPoetry(['install']);
    console.log(chalk`\n  {bgBlue poetry.lock} updated.\n`);
  }
}

export default async function (
  tree: Tree,
  options: PoetryProjectGeneratorSchema
) {
  await checkPoetryExecutable();

  const normalizedOptions = normalizeOptions(tree, options);

  const targets: ProjectConfiguration['targets'] = {
    lock: {
      executor: '@pymonorepo/nx-python:run-commands',
      options: {
        command: 'poetry lock --no-update',
        cwd: normalizedOptions.projectRoot,
      },
    },
    add: {
      executor: '@pymonorepo/nx-python:add',
      options: {},
    },
    update: {
      executor: '@pymonorepo/nx-python:update',
      options: {},
    },
    remove: {
      executor: '@pymonorepo/nx-python:remove',
      options: {},
    },
    build: {
      executor: '@pymonorepo/nx-python:build',
      outputs: ['{projectRoot}/dist'],
      options: {
        outputPath: `${normalizedOptions.projectRoot}/dist`,
        publish: normalizedOptions.publishable,
        lockedVersions: normalizedOptions.buildLockedVersions,
        bundleLocalDependencies: normalizedOptions.buildBundleLocalDependencies,
      },
    },
    install: {
      executor: '@pymonorepo/nx-python:install',
      options: {
        silent: false,
        args: '',
        cacheDir: `.cache/pypoetry`,
        verbose: false,
        debug: false,
      },
    },
  };

  if (options.linter === 'flake8') {
    targets.lint = {
      executor: '@pymonorepo/nx-python:flake8',
      outputs: [
        `{workspaceRoot}/reports/${normalizedOptions.projectRoot}/pylint.txt`,
      ],
      options: {
        outputFile: `reports/${normalizedOptions.projectRoot}/pylint.txt`,
      },
    };
  }

  if (options.unitTestRunner === 'pytest') {
    targets.test = {
      executor: '@pymonorepo/nx-python:run-commands',
      outputs: [
        `{workspaceRoot}/reports/${normalizedOptions.projectRoot}/unittests`,
        `{workspaceRoot}/coverage/${normalizedOptions.projectRoot}`,
      ],
      options: {
        command: `poetry run pytest tests/`,
        cwd: normalizedOptions.projectRoot,
      },
    };
  }

  addProjectConfiguration(tree, normalizedOptions.projectName, {
    root: normalizedOptions.projectRoot,
    projectType: normalizedOptions.projectType,
    sourceRoot: `${normalizedOptions.projectRoot}/${normalizedOptions.moduleName}`,
    targets,
    tags: normalizedOptions.parsedTags,
  });
  addFiles(tree, normalizedOptions);
  updateDevDependenciesProject(tree, normalizedOptions);
  updateRootPyprojectToml(tree, normalizedOptions);
  await formatFiles(tree);

  return () => {
    updateRootPoetryLock(tree);
  };
}
