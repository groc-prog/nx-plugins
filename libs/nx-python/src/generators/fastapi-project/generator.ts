import type { FastAPIProjectGeneratorSchema } from './schema.js';

import * as poetryGenerator from '../poetry-project/generator';
import path from 'path';
import { set } from 'lodash';
import {
  Tree,
  formatFiles,
  generateFiles,
  installPackagesTask,
  names,
  workspaceLayout,
  updateProjectConfiguration,
  readProjectConfiguration,
} from '@nx/devkit';

export default async function generator(tree: Tree, schema: FastAPIProjectGeneratorSchema) {
  await poetryGenerator.default(tree, {
    ...schema,
    type: 'application',
  });

  const projectName = names(schema.name).fileName;
  const moduleName = projectName.replace('-', '_');

  generateFiles(tree, path.join(__dirname, 'files', 'base'), path.join(workspaceLayout().appsDir, projectName), {
    ...schema,
    projectName,
    moduleName,
  });

  if (schema.includeDockerFile)
    generateFiles(tree, path.join(__dirname, 'files', 'docker'), path.join(workspaceLayout().appsDir, projectName), {
      ...schema,
      projectName,
      moduleName,
    });

  const projectConfiguration = readProjectConfiguration(tree, projectName);
  set(projectConfiguration, 'targets.dev', {
    executor: '@nx-python-poetry/nx-python:dev',
    options: {},
  });
  updateProjectConfiguration(tree, projectName, projectConfiguration);

  await formatFiles(tree);
  return () => {
    installPackagesTask(tree);
  };
}
