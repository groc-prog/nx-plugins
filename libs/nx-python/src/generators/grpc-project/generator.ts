import type { Tree } from '@nx/devkit';
import {
  formatFiles,
  generateFiles,
  installPackagesTask,
  names,
  readProjectConfiguration,
  updateProjectConfiguration,
  workspaceLayout,
} from '@nx/devkit';
import { set } from 'lodash';
import toml from '@iarna/toml';
import path from 'path';

import type { GRPCProjectGeneratorSchema } from './schema';
import { PyProjectToml } from '../../utils/poetry';
import * as poetryGenerator from '../poetry-project/generator';

export default async function generator(tree: Tree, schema: GRPCProjectGeneratorSchema) {
  await poetryGenerator.default(tree, {
    ...schema,
    type: 'application',
  });

  const projectName = names(schema.name).fileName;
  const moduleName = projectName.replace('-', '_');
  const projectConfiguration = readProjectConfiguration(tree, projectName);

  generateFiles(tree, path.join(__dirname, 'files', 'base'), path.join(workspaceLayout().appsDir, projectName), {
    ...schema,
    projectName,
    moduleName,
  });

  const projectTomlConfig = tree.read(path.join(workspaceLayout().appsDir, projectName, 'pyproject.toml'), 'utf-8');
  const projectTomlData = toml.parse(projectTomlConfig) as PyProjectToml;

  set(projectTomlData, 'tool.nx', { port: schema.port, kind: 'grpc', host: schema.host });
  set(projectTomlData, 'tool.poetry.dependencies.grpcio', '*');
  set(projectTomlData, 'tool.poetry.dependencies.grpcio-reflection', '*');
  set(projectTomlData, 'tool.poetry.group.dev.dependencies.watchdog', '*');
  tree.write(path.join(workspaceLayout().appsDir, projectName, 'pyproject.toml'), toml.stringify(projectTomlData));

  if (schema.includeDockerFile) {
    generateFiles(tree, path.join(__dirname, 'files', 'docker'), path.join(workspaceLayout().appsDir, projectName), {
      ...schema,
      projectName,
      moduleName,
      appsDir: workspaceLayout().appsDir,
    });
  }

  set(projectConfiguration, 'targets.dev', {
    executor: '@nx-python-poetry/nx-python:dev',
    options: {},
  });
  updateProjectConfiguration(tree, projectName, projectConfiguration);

  await formatFiles(tree);
  return () => {
    installPackagesTask(tree);
  };
}
