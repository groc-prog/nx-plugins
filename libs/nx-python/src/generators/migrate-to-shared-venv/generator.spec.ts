import { spawnSyncMock } from '../../utils/mocks/cross-spawn.mock';
import * as poetryUtils from '../../executors/utils/poetry';
import { Tree } from '@nx/devkit';
import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import generator from './generator';
import projectGenerator from '../project/generator';

describe('nx-python migrate-shared-venv generator', () => {
  let checkPoetryExecutableMock: jest.SpyInstance;
  let appTree: Tree;

  beforeEach(() => {
    appTree = createTreeWithEmptyWorkspace({
      layout: 'apps-libs',
    });
    checkPoetryExecutableMock = jest.spyOn(
      poetryUtils,
      'checkPoetryExecutable'
    );
    checkPoetryExecutableMock.mockResolvedValue(undefined);
    spawnSyncMock.mockReturnValue({ status: 0 });
  });

  it('should throw an exception when the poetry is not installed', async () => {
    checkPoetryExecutableMock.mockRejectedValue(new Error('poetry not found'));

    expect(
      generator(appTree, {
        moveDevDependencies: true,
        pyenvPythonVersion: '3.8.11',
        pyprojectPythonDependency: '>=3.8,<3.10',
        autoActivate: false,
      })
    ).rejects.toThrow('poetry not found');

    expect(checkPoetryExecutableMock).toHaveBeenCalled();
  });

  it('should migrate an isolate venv to shared venv', async () => {
    await projectGenerator(appTree, {
      name: 'proj1',
      type: 'application',
      publishable: true,
      customSource: false,
      addDevDependencies: true,
      moduleName: 'proj1',
      packageName: 'proj1',
      buildLockedVersions: true,
      buildBundleLocalDependencies: true,
      pyenvPythonVersion: '3.8.11',
      pyprojectPythonDependency: '>=3.8,<3.10',
      toxEnvlist: 'py38',
    });

    const task = await generator(appTree, {
      moveDevDependencies: true,
      pyenvPythonVersion: '3.8.11',
      pyprojectPythonDependency: '>=3.8,<3.10',
      autoActivate: false,
    });
    task();

    expect(checkPoetryExecutableMock).toHaveBeenCalled();
    expect(
      appTree.read('apps/proj1/pyproject.toml', 'utf-8')
    ).toMatchSnapshot();
    expect(appTree.read('pyproject.toml', 'utf-8')).toMatchSnapshot();
    expect(appTree.read('.python-version', 'utf-8')).toMatchSnapshot();
    expect(spawnSyncMock).toHaveBeenNthCalledWith(
      1,
      'poetry',
      ['lock', '--no-update'],
      { cwd: 'apps/proj1', shell: false, stdio: 'inherit' }
    );
    expect(spawnSyncMock).toHaveBeenNthCalledWith(2, 'poetry', ['install'], {
      shell: false,
      stdio: 'inherit',
    });
  });

  it('should migrate an isolate venv to shared venv project without dev dependencies', async () => {
    await projectGenerator(appTree, {
      name: 'proj1',
      type: 'application',
      publishable: true,
      customSource: false,
      addDevDependencies: false,
      moduleName: 'proj1',
      packageName: 'proj1',
      buildLockedVersions: true,
      buildBundleLocalDependencies: true,
      pyenvPythonVersion: '3.8.11',
      pyprojectPythonDependency: '>=3.8,<3.10',
      toxEnvlist: 'py38',
    });

    const task = await generator(appTree, {
      moveDevDependencies: true,
      pyenvPythonVersion: '3.8.11',
      pyprojectPythonDependency: '>=3.8,<3.10',
      autoActivate: false,
    });
    task();

    expect(checkPoetryExecutableMock).toHaveBeenCalled();
    expect(
      appTree.read('apps/proj1/pyproject.toml', 'utf-8')
    ).toMatchSnapshot();
    expect(appTree.read('pyproject.toml', 'utf-8')).toMatchSnapshot();
    expect(appTree.read('.python-version', 'utf-8')).toMatchSnapshot();
    expect(spawnSyncMock).toHaveBeenNthCalledWith(
      1,
      'poetry',
      ['lock', '--no-update'],
      { cwd: 'apps/proj1', shell: false, stdio: 'inherit' }
    );
    expect(spawnSyncMock).toHaveBeenNthCalledWith(2, 'poetry', ['install'], {
      shell: false,
      stdio: 'inherit',
    });
  });

  it('should migrate an isolate venv to shared venv with auto activate enabled', async () => {
    await projectGenerator(appTree, {
      name: 'proj1',
      type: 'application',
      publishable: true,
      customSource: false,
      addDevDependencies: true,
      moduleName: 'proj1',
      packageName: 'proj1',
      buildLockedVersions: true,
      buildBundleLocalDependencies: true,
      pyenvPythonVersion: '3.8.11',
      pyprojectPythonDependency: '>=3.8,<3.10',
      toxEnvlist: 'py38',
    });

    const task = await generator(appTree, {
      moveDevDependencies: true,
      pyenvPythonVersion: '3.8.11',
      pyprojectPythonDependency: '>=3.8,<3.10',
      autoActivate: true,
    });
    task();

    expect(checkPoetryExecutableMock).toHaveBeenCalled();
    expect(appTree.read('pyproject.toml', 'utf-8')).toMatchSnapshot();
  });
});
