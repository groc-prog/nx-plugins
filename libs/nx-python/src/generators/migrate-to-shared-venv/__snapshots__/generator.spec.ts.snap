// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`nx-python migrate-shared-venv generator should migrate an isolate venv to shared venv 1`] = `
"[tool.coverage.run]
branch = true
source = [ "proj1" ]

[tool.coverage.report]
exclude_lines = [ "if TYPE_CHECKING:" ]
show_missing = true

[tool.pytest.ini_options]
addopts = "--cov --cov-fail-under=100 --cov-report html:'../../coverage/apps/proj1/html' --cov-report xml:'../../coverage/apps/proj1/coverage.xml' --junitxml='../../reports/apps/proj1/unittests/junit.xml' --html='../../reports/apps/proj1/unittests/html/index.html'"

[tool.poetry]
name = "proj1"
version = "1.0.0"
description = ""
authors = [ ]
license = "Proprietary"
readme = "README.md"
group = { }

  [[tool.poetry.packages]]
  include = "proj1"

  [tool.poetry.dependencies]
  python = ">=3.8,<3.10"

[build-system]
requires = [ "poetry-core==1.1.0" ]
build-backend = "poetry.core.masonry.api"
"
`;

exports[`nx-python migrate-shared-venv generator should migrate an isolate venv to shared venv 2`] = `
"[tool.poetry]
name = "@nxlv/nx-plugins"
version = "1.0.0"
description = ""
authors = [ ]
license = "Proprietary"
readme = "README.md"

  [tool.poetry.dependencies]
  python = ">=3.8,<3.10"

    [tool.poetry.dependencies.proj1]
    path = "apps/proj1"
    develop = true

[tool.poetry.group.dev.dependencies]
flake8 = "4.0.1"
flake8-isort = "4.1.1"
flake8-print = "5.0.0"
flake8-pytest-style = "1.6.0"
flake8-docstrings = "1.6.0"
flake8-type-checking = "2.0.6"
autopep8 = "1.5.7"
pytest = "7.1.2"
pytest-env = "0.6.2"
pytest-cov = "3.0.0"
pytest-html = "3.1.1"
pytest-sugar = "0.9.5"
tomli = "1.2.2"
tox = "3.23.1"

[build-system]
requires = [ "poetry-core==1.1.0" ]
build-backend = "poetry.core.masonry.api"
"
`;

exports[`nx-python migrate-shared-venv generator should migrate an isolate venv to shared venv 3`] = `
"3.8.11
"
`;

exports[`nx-python migrate-shared-venv generator should migrate an isolate venv to shared venv project without dev dependencies 1`] = `
"[tool.coverage.run]
branch = true
source = [ "proj1" ]

[tool.coverage.report]
exclude_lines = [ "if TYPE_CHECKING:" ]
show_missing = true

[tool.pytest.ini_options]
addopts = "--cov --cov-fail-under=100 --cov-report html:'../../coverage/apps/proj1/html' --cov-report xml:'../../coverage/apps/proj1/coverage.xml' --junitxml='../../reports/apps/proj1/unittests/junit.xml' --html='../../reports/apps/proj1/unittests/html/index.html'"

[tool.poetry]
name = "proj1"
version = "1.0.0"
description = ""
authors = [ ]
license = "Proprietary"
readme = "README.md"

  [[tool.poetry.packages]]
  include = "proj1"

  [tool.poetry.dependencies]
  python = ">=3.8,<3.10"

[build-system]
requires = [ "poetry-core==1.1.0" ]
build-backend = "poetry.core.masonry.api"
"
`;

exports[`nx-python migrate-shared-venv generator should migrate an isolate venv to shared venv project without dev dependencies 2`] = `
"[tool.poetry]
name = "@nxlv/nx-plugins"
version = "1.0.0"
description = ""
authors = [ ]
license = "Proprietary"
readme = "README.md"

  [tool.poetry.dependencies]
  python = ">=3.8,<3.10"

    [tool.poetry.dependencies.proj1]
    path = "apps/proj1"
    develop = true

[build-system]
requires = [ "poetry-core==1.1.0" ]
build-backend = "poetry.core.masonry.api"
"
`;

exports[`nx-python migrate-shared-venv generator should migrate an isolate venv to shared venv project without dev dependencies 3`] = `
"3.8.11
"
`;

exports[`nx-python migrate-shared-venv generator should migrate an isolate venv to shared venv with auto activate enabled 1`] = `
"[tool.nx]
autoActivate = true

[tool.poetry]
name = "@nxlv/nx-plugins"
version = "1.0.0"
description = ""
authors = [ ]
license = "Proprietary"
readme = "README.md"

  [tool.poetry.dependencies]
  python = ">=3.8,<3.10"

    [tool.poetry.dependencies.proj1]
    path = "apps/proj1"
    develop = true

[tool.poetry.group.dev.dependencies]
flake8 = "4.0.1"
flake8-isort = "4.1.1"
flake8-print = "5.0.0"
flake8-pytest-style = "1.6.0"
flake8-docstrings = "1.6.0"
flake8-type-checking = "2.0.6"
autopep8 = "1.5.7"
pytest = "7.1.2"
pytest-env = "0.6.2"
pytest-cov = "3.0.0"
pytest-html = "3.1.1"
pytest-sugar = "0.9.5"
tomli = "1.2.2"
tox = "3.23.1"

[build-system]
requires = [ "poetry-core==1.1.0" ]
build-backend = "poetry.core.masonry.api"
"
`;
